### PR TITLE
MOD-8172: Fix CI for platforms that don't support Node.js > 16

### PR DESCRIPTION
**Pull request by raz-mon**
GitHub runners don't support Node.js 16 anymore + RediSearch still supports platforms that do not support any newer versions of Node.js = We can't use the `@v3` versions of the `checkout` and `configure-aws-credential` actions, which until now allowed to use Node.js 16.
--> This PR replaces the use of these actions with bash scripts that will run on the corresponding platform (container), such that we do not need to execute any TS code (via the Node.js environment).

### CODE REVIEW

File: .github/workflows/task-build-artifacts.yml [hunk 1]
```diff
@@ -80,20 +78,73 @@ jobs:
...
-          ref: ${{ env.REF }}
-          sparse-checkout-cone-mode: false
-          sparse-checkout: |
-            .install
-            tests/pytests/requirements.txt
-      - name: Setup specific
+        run: |
+          # Checkout script logic
+          perform_checkout_ref() {
+            REPO_URL="https://github.com/RediSearch/RediSearch.git"
```
Comment (oshadmi) [inline .github/workflows/task-build-artifacts.yml:84]: Can we use something generic like
```
REPO_URL="https://github.com/${{ github.repository }}"
```

File: .github/workflows/task-build-artifacts.yml [hunk 2]
```diff
@@ -80,20 +78,73 @@ jobs:
...
+          # Execute the logic based on the detected platform
+          echo "Detected platform: ${{ inputs.container }}"
+          case "${{ inputs.container }}" in
+            ubuntu:bionic | amazonlinux:2 | alpine:3)
+              echo "Installing prerequisites for ${{ inputs.container }}..."
+
+              # Install git (platform-specific package management)
+              case "${{ inputs.container }}" in
+                ubuntu:bionic)
+                  echo "Updating package lists..."
```
Comment (oshadmi) [inline .github/workflows/task-build-artifacts.yml:107]: Echo messages also in other cases?

File: .github/workflows/task-build-artifacts.yml [hunk 3]
```diff
@@ -158,10 +212,20 @@ jobs:
...
-          role-to-assume: ${{ vars.ARTIFACT_UPLOAD_AWS_ROLE_TO_ASSUME }}
-          aws-region: ${{ vars.ARTIFACT_UPLOAD_AWS_REGION }}
+        run: |
+          # Variables from the workflow
+          ROLE_ARN="${{ vars.ARTIFACT_UPLOAD_AWS_ROLE_TO_ASSUME }}"
+          AWS_REGION="${{ vars.ARTIFACT_UPLOAD_AWS_REGION }}"
+          SESSION_NAME="github-actions-session"
+
+          # Assume the AWS role and configure temporary credentials
+          echo "Assuming AWS Role: $ROLE_ARN"
```
Comment (oshadmi) [inline .github/workflows/task-build-artifacts.yml:222]: Do we need such echoing?

File: .github/workflows/task-build-artifacts.yml [hunk 4]
```diff
@@ -171,10 +235,26 @@ jobs:
...
+
+          # Configure AWS CLI with provided credentials and region
+          echo "Configuring AWS CLI with access keys..."
+          aws configure set aws_access_key_id "$AWS_ACCESS_KEY_ID"
+          aws configure set aws_secret_access_key "$AWS_SECRET_ACCESS_KEY"
+          aws configure set region "$AWS_REGION"
+
+          # Verify configuration
+          echo "Verifying AWS CLI configuration..."
+          aws sts get-caller-identity
```
Comment (oshadmi) [inline .github/workflows/task-build-artifacts.yml:258]: Do we need such output?

File: .github/workflows/task-build-artifacts.yml (hunk 2 above)
Comment (raz-mon) [reply .github/workflows/task-build-artifacts.yml:107]: One print before is good enough, removing this one.

File: .github/workflows/task-test.yml [hunk 5]
```diff
@@ -80,26 +75,61 @@ jobs:
...
-        uses: actions/checkout@v3
-        with:
-          path: setup
-          sparse-checkout-cone-mode: false
-          sparse-checkout: |
-            .install
-            tests/pytests/requirements.*
-      - name: Setup specific
+        run: |
+          # Execute the logic based on the detected platform
```
Comment (oshadmi) [inline .github/workflows/task-test.yml:79]: If this snippet is repeated, consider putting it in a bash script and reuse it?

File: .github/workflows/task-build-artifacts.yml (hunk 3 above)
Comment (raz-mon) [reply .github/workflows/task-build-artifacts.yml:222]: No, removed

File: .github/workflows/task-build-artifacts.yml (hunk 4 above)
Comment (raz-mon) [reply .github/workflows/task-build-artifacts.yml:258]: The print?
Don't see how it hurts

File: .github/workflows/task-build-artifacts.yml [hunk 6]
```diff
@@ -56,14 +59,9 @@ jobs:
       - name: Check if node20 is Supported
         id: node20 # TODO: Remove this when node20 is supported on all platforms, or when we drop support for theses platforms
         run: |
-          for platform in ubuntu:bionic centos:7 amazonlinux:2 alpine:3; do
```
Comment (kei-nan) [inline .github/workflows/task-build-artifacts.yml:59]: Why was centos:7 removed?

File: .github/workflows/task-test.yml (hunk 5 above)
Comment (kei-nan) [reply .github/workflows/task-test.yml:79]: You could use a separate repo to house your custom actions and expose them as a github action.

File: .github/workflows/task-build-artifacts.yml (hunk 6 above)
Comment (raz-mon) [reply .github/workflows/task-build-artifacts.yml:59]: Was removed from our CI [here](https://github.com/RediSearch/RediSearch/pull/4675) (in master - if it still exists in the version branches it will not be removed there).

File: .github/workflows/task-test.yml [hunk 7]
```diff
@@ -172,11 +201,7 @@ jobs:
         if: >
           steps.node20.outputs.supported == 'false' &&
           (steps.unit_tests.outcome == 'failure' || steps.standalone_tests.outcome == 'failure' || steps.coordinator_tests.outcome == 'failure')
-        uses: actions/upload-artifact@v3
-        with:
-          name: Test Logs ${{ steps.artifact-names.outputs.name }}
-          path: tests/**/logs/*.log*
-          if-no-files-found: ignore
+        run: echo "Currently not available..."
```
Comment (oshadmi) [inline .github/workflows/task-test.yml:204]: Can do it in a separate PR (maybe even just a `grep -C` for some `[FAIL]` etc. tokens

**Review by oshadmi — DISMISSED** [review]

File: .github/workflows/task-build-artifacts.yml (hunk 6 above)
Comment (raz-mon) [reply .github/workflows/task-build-artifacts.yml:59]: I see it is not supported on any of the version branches as well (which makes sense..)

File: .github/workflows/task-test.yml [hunk 8]
```diff
@@ -80,26 +75,59 @@ jobs:
...
-        with:
-          path: setup
-          sparse-checkout-cone-mode: false
-          sparse-checkout: |
-            .install
-            tests/pytests/requirements.*
-      - name: Setup specific
+        run: |
+          # Execute the logic based on the detected platform
+          echo "Detected platform: ${{ inputs.container }}"
```
Comment (kei-nan) [inline .github/workflows/task-test.yml:80]: Can all of these bash scripts move to separate yml files?
It might also allow some code reuse.

File: .github/workflows/task-build-artifacts.yml (hunk 6 above)
Comment (GuyAv46) [reply .github/workflows/task-build-artifacts.yml:59]: it should be supported on 2.6, but it was listed here just because there was no good reason to remove it (has no effect)

File: .github/workflows/task-test.yml (hunk 8 above)
Comment (raz-mon) [reply .github/workflows/task-test.yml:80]: This is before we have our repo cloned, so I don't think so. At this point the machine only has this flow in hand.

Comment (raz-mon) [reply .github/workflows/task-test.yml:80]: Maybe we can hold scripts in the repo and use them like variables

File: .github/workflows/task-test.yml [hunk 9]
```diff
@@ -172,11 +201,7 @@ jobs:
         if: >
```
Comment (GuyAv46) [inline .github/workflows/task-test.yml:201]: consider updating the comment here

File: .github/workflows/task-test.yml [hunk 10]
```diff
@@ -80,26 +75,59 @@ jobs:
...
         if: steps.node20.outputs.supported == 'true'
         uses: actions/checkout@v4
         with:
           submodules: recursive
       - name: Full checkout (node20 not supported)
         if: steps.node20.outputs.supported == 'false'
-        uses: actions/checkout@v3
-        with:
-          submodules: recursive
+        run: echo "Done already earlier."
```
Comment (GuyAv46) [inline .github/workflows/task-test.yml:130]: Consider grouping all the node-20 steps together and all the non-node-20 together. I think it will make more sense than having these dummy steps

File: .github/workflows/task-build-artifacts.yml (hunk 6 above)
Comment (raz-mon) [reply .github/workflows/task-build-artifacts.yml:59]: Right, 2.6 as well.
It's confusing w.r.t to whether it is supported on latest or not, I prefer it not to appear where it is not supported

File: .github/workflows/task-test.yml (hunk 9 above)
Comment (raz-mon) [reply .github/workflows/task-test.yml:201]: You mean to remove the failure condition?

**Review by GuyAv46 — APPROVED** [review]

File: .github/workflows/task-test.yml (hunk 8 above)
Comment (kei-nan) [reply .github/workflows/task-test.yml:80]: How does this script get called?
Isn't it used like this?
```
uses: ./.github/workflows/task-test.yml
```

Comment (raz-mon) [reply .github/workflows/task-test.yml:80]: Now I see your meaning - sounds good will try to add it.
